dvc/repo/metrics/show.py

```diff
@@ -80,71 +129,23 @@ def show(
     for branch in self.scm.brancher(
         all_branches=all_branches, all_tags=all_tags
     ):
-        astages = self.active_stages()
-        outs = [out for stage in astages for out in stage.outs]
-
-        if path:
-            outs = find_output_by_path(
-                self, path, outs=outs, recursive=recursive
-            )
-
-        metrics = filter(lambda o: o.metric, outs)
-        stages = set()
-        entries = []
-
-        for o in metrics:
-            if not typ and isinstance(o.metric, dict):
-                t = o.metric.get(o.PARAM_METRIC_TYPE, typ)
-                x = o.metric.get(o.PARAM_METRIC_XPATH, xpath)
-            else:
-                t = typ
-                x = xpath
-            entries.append((o.path, t, x))
-            stages.add(o.stage.path)
+        entries = _collect_metrics(self, path, recursive, typ, xpath)
 
         if path and not entries:
             if os.path.isdir(path):
                 logger.warning(
                     "Path '{path}' is a directory. "
                     "Consider running with '-R'.".format(path=path)
                 )
-                return {}
-
-        for fname, t, x in entries:
-            if stages:
-                for stage in stages:
-                    self.checkout(stage, force=True)
-
-            rel = os.path.relpath(fname)
-            metric = _read_metric(fname, typ=t, xpath=x)
-            if not metric:
                 continue
```

Comment:
`return {}` probably make more sense here, there is no reason to continue if path provided is a directory

Reply:
@shcheklein It might not be a directory on another branch though, so we can't return here.

Reply:
we can of course, it's just a matter of semantics we want to support in this and similar edge cases. It's not clear how should we behave when there is a mix of directories and files across branches. The same for `-R file` for example. It look pretty strange when you run it in the branch where `file` is not a directory. Should we allow it or fail safely with warning? Probably you are even right that we should allow all of this, but it looks a bit weird :)

Reply:
I agree, we should support both. Re-did our output search so that it considers `-R name` valid both for `name` dirs and `name` files, which is how it works right now. Removed that warning since it is not necessary.